util/label/label.rs

```diff
@@ -0,0 +1,332 @@
+    if !input.starts_with("@") {
+        return Ok((input, None));
+    }
+
+    let slash_pos = input
+        .find("//")
+        .ok_or(err(label, "labels with repository must contain //."))?;
+    let repository_name = &input[1..slash_pos];
+    if repository_name.is_empty() {
+        return Ok((&input[1..], None));
+    }
+    if !repository_name
+        .chars()
+        .next()
+        .unwrap()
+        .is_ascii_alphabetic()
+    {
+        return Err(LabelError(err(
+            label,
+            "workspace names must start with a letter.",
+        )));
+    }
+    if !repository_name
+        .chars()
+        .all(|c| c.is_ascii_alphanumeric() || c == '-' || c == '_' || c == '.')
+    {
+        return Err(LabelError(err(
+            label,
+            "workspace names \
+                may contain only A-Z, a-z, 0-9, '-', '_', and '.'.",
+        )));
+    }
+    Ok((&input[slash_pos..], Some(repository_name)))
+}
+
+fn consume_package_name<'s>(
+    input: &'s str,
+    label: &'s str,
+) -> Result<(&'s str, Option<&'s str>, Option<&'s str>)> {
+    let colon_pos = input.find(":");
+    let start_pos;
+    let mut is_absolute = false;
+    if input.starts_with("//") {
+        start_pos = 2;
+        is_absolute = true;
+    } else {
+        start_pos = 0;
+        if colon_pos.is_none() {
+            if input.contains("//") {
+                return Err(LabelError(err(
+                    label,
+                    "workspace name must start with '@'.",
+                )));
+            }
+            return Ok((input, None, None));
+        }
+    };
+
+    let (package_name, rest) = match colon_pos {
+        Some(colon_pos) => (&input[start_pos..colon_pos], &input[colon_pos..]),
+        None => (&input[start_pos..], ""),
+    };
+    if package_name.is_empty() {
+        return Ok((rest, None, None));
+    }
+    if package_name.contains("//") {
+        return Err(LabelError(err(
+            label,
+            "workspace name must start with '@'.",
+        )));
+    }
+    let implicit_target_name = if is_absolute {
+        match package_name.rfind("/") {
+            Some(pos) => Some(&package_name[pos..]),
+            None => Some(package_name),
+        }
+    } else {
+        None
+    };
+
+    if !package_name.chars().all(|c| {
+        c.is_ascii_alphanumeric()
+            || c == '/'
+            || c == '-'
+            || c == '.'
+            || c == ' '
+            || c == '$'
+            || c == '('
+            || c == ')'
+            || c == '_'
+    }) {
+        return Err(LabelError(err(
+            label,
+            "package names may contain only A-Z, \
+        a-z, 0-9, '/', '-', '.', ' ', '$', '(', ')' and '_'.",
+        )));
+    }
+    if package_name.ends_with("/") {
+        return Err(LabelError(err(
+            label,
+            "package names may not end with '/'.",
+        )));
+    }
+
+    Ok((rest, Some(package_name), implicit_target_name))
```

Comment:
Looks like we end up using either rest or implicit_target_name later on, maybe make this function return one of them instead?